### PR TITLE
(closes 1587) fast debug string method

### LINKED ISSUE #1424: [psyir] add functionality for tree comparison
In `ArrayMixin._matching_access()` we need to compare two references to see whether they are equivalent. At the moment we resort to using the visitor to do this but this is not a nice solution. We should be able to compare two PSyIR fragments 'natively'.

### PR DESCRIPTION
**Pull request by sergisiso**
A new backend and method to generate Fortran-like code quickly by not lowering nor deepcopying the tree.

Maybe it also closes #1424?

### CODE REVIEW

**Comment by sergisiso** [comment]
@arporter or @rupertford  This is ready for a first review

**Comment by arporter** [comment]
Hi @sergisiso, please could you fix the coverage issues and then I'll take a look.

**Comment by sergisiso** [comment]
@arporter Done now (it was dummy lines on the src/psyclone/test)

File: src/psyclone/tests/psyir/backend/sympy_writer_test.py [hunk 1]
```diff
@@ -47,14 +48,29 @@
...
     '''Test that the constructor accepts an optional dictionary.
     '''
     sympy_writer = SymPyWriter({'some': 'symbol'})
     assert sympy_writer._sympy_type_map['some'] == 'symbol'
     # Also test that not specifying a type map as argument works:
     sympy_writer = SymPyWriter()
     assert sympy_writer._sympy_type_map == {}
+    assert sympy_writer._DISABLE_LOWERING is True
+
+    def error(_):
```
Comment (arporter) [inline src/psyclone/tests/psyir/backend/sympy_writer_test.py:61]: I think from hereon should be a new test - maybe test_lowering_disabled or something.

**Review by arporter — CHANGES_REQUESTED** [review]
Nice work Sergi. The new class makes things a lot simpler.
Just a few very minor things to address.
Tests and examples are all fine with compilation.
Docs all build OK (and have been updated).
sympy_writer still has a reference to #1587.

File: src/psyclone/psyir/backend/debug_writer.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    lowering the higher level concepts to language constructs. This special
+    backend can do it because instead of failing for an unknown construct it
+    will just print the name of the higher level Node.
+    The resulting code will not be compilable but it will be usable for
+    generating debug information much faster than the FortranWriter because
+    it does not need to lower the nodes and therefore it does not need to
+    deepcopy the tree.
+
+    '''
+    _DISABLE_LOWERING = True
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/backend/debug_writer.py:54]: Done

File: src/psyclone/psyir/backend/debug_writer.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        super().__init__(check_global_constraints=False)
+
+    def node_node(self, node):
+        ''' The DebugWriter must never fail with unrecognized nodes, this
+        generic visitor will capture any Node that the super Fortran backend
+        did not catch and output it as "< Node string >".
+
+        :param node: the PSyIR node to translate.
+        :type node: :py:class:`psyclone.psyir.node.Node`
+
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/backend/debug_writer.py:66]: Done

File: src/psyclone/psyir/backend/sympy_writer.py [hunk 4]
```diff
@@ -67,6 +68,7 @@ class SymPyWriter(FortranWriter):
         :py:meth:`psyclone.core.sympy_writer.create_type_map`.
     :type type_map: dict of str:Sympy-data-type values
     '''
+    _DISABLE_LOWERING = True
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/backend/sympy_writer.py:71]: Done

File: src/psyclone/psyir/backend/visitor.py [hunk 5]
```diff
@@ -80,6 +80,7 @@ class PSyIRVisitor():
 
     '''
     _COMMENT_PREFIX = None
+    _DISABLE_LOWERING = False
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/backend/visitor.py:83]: Done

File: src/psyclone/psyir/backend/visitor.py [hunk 6]
```diff
@@ -161,15 +162,18 @@ def __call__(self, node):
                 f"The PSyIR visitor functor method only accepts a PSyIR Node "
                 f"as argument, but found '{type(node).__name__}'.")
 
+        if self._DISABLE_LOWERING:
+            return self._visit(node)
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/backend/visitor.py:166]: Done

File: src/psyclone/psyir/nodes/node.py [hunk 7]
```diff
@@ -1435,6 +1435,18 @@ def validate_global_constraints(self):
 
         '''
 
+    def debug_string(self):
+        ''' Generates a Fortran-like output representation but without
+        lowering high-level nodes. This is fast to generate because it
+        doesn't deepcopy the tree like the Language backends and its
+        output, although not compilable, is readable for error messages.
+
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/nodes/node.py:1443]: Done

File: src/psyclone/tests/psyad/transformations/test_adjoint_trans.py [hunk 8]
```diff
@@ -83,14 +83,3 @@ def name(self):
...
             "'NoneType'." in str(info.value))
-
-    active_vars = [DataSymbol("x", REAL_TYPE)]
-    with pytest.raises(TypeError) as info:
-        DummyTrans(active_vars, None)
-    assert ("The writer argument should be a PSyIRVisitor but found "
-            "'NoneType'." in str(info.value))
-
-    writer = SIRWriter()
-    dummy_trans = DummyTrans(active_vars, writer)
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/psyad/transformations/test_adjoint_trans.py:94]: ups.. reverted

File: src/psyclone/tests/psyir/backend/debug_writer_test.py [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+    real, intent(in),  dimension(:,:) :: p, u
+
+    CU(I,J) = 0.5d0*(P(i+1,J)+P(I,J))*U(I,J)
+
+  end subroutine example_code
+end module example_mod
+'''
+
+
+def test_debug_writer_fotran(fortran_reader, fortran_writer):
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/psyir/backend/debug_writer_test.py:62]: Done

File: src/psyclone/tests/psyir/backend/debug_writer_test.py [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+
+    !$omp parallel default(shared)
+    cu(i,j) = < MyDSLDataNode[] > * u(i,j)
+    !$omp end parallel
+    < MyDSLStatement[] >
+
+  end subroutine example_code
+
+end module example_mod\n'''
+
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/psyir/backend/debug_writer_test.py:139]: Removed

File: src/psyclone/tests/psyir/backend/sympy_writer_test.py [hunk 11]
```diff
@@ -47,14 +48,29 @@
...
+        ''' Just produce and error '''
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/psyir/backend/sympy_writer_test.py:62]: Done

File: src/psyclone/tests/psyir/backend/sympy_writer_test.py (hunk 1 above)
Comment (sergisiso) [reply src/psyclone/tests/psyir/backend/sympy_writer_test.py:61]: Done

File: src/psyclone/psyir/nodes/array_mixin.py [hunk 12]
```diff
@@ -373,23 +373,10 @@ def is_same_array(self, node):
...
-            # TODO #1424. We need to be able to compare PSyIR fragments
-            # natively rather than using a visitor. We use the `_visit` method
-            # directly here so as to avoid the deep-copying of the complete
-            # tree which is performed when using fwriter(). (This operation
-            # can become very, very costly for large trees.)
-            # pylint: disable=protected-access
-            if ("".join(fwriter._visit(idx) for idx in indices[0]) !=
-                    "".join(fwriter._visit(idx) for idx in indices[1])):
+        for idx1, idx2 in zip(self_indices[:depth], node_indices[:depth]):
+            if idx1 != idx2:
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/nodes/array_mixin.py:379]: Unless I am misunderstanding what you mean, indeed operation equality is implemented and will look for the same operator and subtree in both sides of the comparison.

Symbolic equality is not considered here but we could expand this in the future ("2" == "1+1"). But this also didn't work in the string comparison version before.

**Comment by sergisiso** [comment]
@arporter This is ready for the next review

File: src/psyclone/psyir/nodes/array_mixin.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/nodes/array_mixin.py:379]: Yes, I think that was what I meant so that's good :-)

**Review by arporter — APPROVED** [review]
All requested changes have been made.
Tests and examples are still all fine.
Ref. guide is still fine.
Mention of #1587 has been removed.
Will proceed to merge.

**Pull request merged**